cmd/snap/cmd_run.go

```diff
@@ -954,3 +967,88 @@ func (x *cmdRun) runSnapConfine(info *snap.Info, securityTag, snapApp, hook stri
 		return syscallExec(cmd[0], cmd, env)
 	}
 }
+
+func randomUUID() (string, error) {
+	uuidBytes, err := ioutil.ReadFile("/proc/sys/kernel/random/uuid")
+	return strings.TrimSpace(string(uuidBytes)), err
+}
+
+var createTransientScope = func(securityTag string) error {
+	if !features.RefreshAppAwareness.IsEnabled() {
+		return nil
+	}
+
+	// The scope is created with a DBus call to systemd running either on
+	// system or session bus, depending on if we are starting a program as root
+	// or as a regular user.
+	var conn *dbus.Conn
+	var err error
+	if os.Getuid() == 0 {
+		conn, err = dbus.SystemBus()
+	} else {
+		conn, err = dbus.SessionBus()
+	}
+	if err != nil {
+		return err
+	}
+
+	// The property and auxUnit types are not documented but can be traced from
+	// systemd source code. Systemd defines the signature of StartTransientUnit
+	// as "ssa(sv)a(sa(sv))". The signature can be decomposed as follows:
+	//
+	// unitName string // name of the unit to start
+	// jobMode string  // corresponds to --job-mode= (see systemctl(1) manual page)
+	// properties []struct{
+	//   Name string
+	//   Value interface{}
+	// } // properties describe properties of the started unit
+	// auxUnits []struct {
+	//   Name string
+	//   Properties []struct{
+	//   	Name string
+	//   	Value interface{}
+	//	 }
+	// } // auxUnits describe any additional units to define.
+	type property struct {
+		Name  string
+		Value interface{}
+	}
+	type auxUnit struct {
+		Name  string
+		Props []property
+	}
+
+	// We ask the kernel for a random UUID. We need one because each transient
+	// scope needs a unique name. The unique name is comprosed of said UUID and
+	// the snap security tag.
+	uuid, err := randomUUID()
+	if err != nil {
+		return err
+	}
+	// Instead of enforcing uniqueness we could join an existing scope but this has some limitations:
+	// - the originally started scope must be marked as a delegate, with all consequences.
+	// - the method AttachProcessesToUnit is unavailable on Ubuntu 16.04
+	unitName := fmt.Sprintf("snap.%s.%s.scope", uuid, strings.TrimPrefix(securityTag, "snap."))
+	mode := "fail"
+	properties := []property{{"PIDs", []uint{uint(os.Getpid())}}}
+	aux := []auxUnit(nil)
+	systemd := conn.Object("org.freedesktop.systemd1", "/org/freedesktop/systemd1")
+	call := systemd.Call(
```

Comment:
Out of curiosity, how long does this call typically take?

Reply:
I'll work on measuring this in a reproducible way but when I was doing casual measurements it just doesn't register. It is in no way slower than previous code, that I could measure.

Reply:
I used bustle to capture the time required for activation of a simple "bin true" snap

<img width="707" alt="bustle" src="https://user-images.githubusercontent.com/784262/74839702-02428d00-5326-11ea-8920-d6f6d0f9ba81.png">

zyga@bionic:~$ time snapd-hacker-toolbelt.busybox true

I also collected timings from running that in the shell:
```
real	0m0.050s
user	0m0.042s
sys	0m0.005s
```

As you can see the activation takes a millisecond. Garbage collection of the dead cgroup takes some more but doesn't impact responsiveness of the system.

Reply:
With the feature disabled:
```
zyga@bionic:~$ time snapd-hacker-toolbelt.busybox true

real	0m0.049s
user	0m0.030s
sys	0m0.020s
```